csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -215,6 +216,706 @@ class CircularBufferLoopCloner : public kir::IrVisitor {
   const std::unordered_set<Expr*>& exclude_;
 };
 
+// Description:
+// Replicates circular buffer loops for Prologue, Main, and
+// Epilogue. Prologue only copies the load expressions of circular
+// buffered tensors, whereas Epilogue does any expression other than
+// the loads. Main copies everything. The pre-prologue and post-epilogue loops
+// are created separately by createCpAsyncBulkFixtures.
+//
+// Loop Structure Overview:
+// Pre-prologue loop:
+// - Allocate shared memory for mbarriers and mbarrier tokens
+// - Initialize mbarrier for all stages
+//
+// Prologue loop:
+// - if selected_thread:
+//   - Issue cp async bulks for all but last stage
+//
+// Main loop:
+// - if selected_thread:
+//   - Issue next cp async bulk for available stage
+// - All threads wait until tma operation arrives
+// - Copy body without
+//   - shared memory allocations
+//   - mbarrier_init exprs
+//   - mbarrier_inval exprs
+//
+// Epilogue loop:
+// - All threads wait until tma operation arrives
+// - Copy body without
+//   - shared memory allocations
+//   - issuing cp async bulk operations
+//   - mbarrier_init exprs
+//   - mbarrier_inval exprs
+//
+// Post-epilogue loop:
+//  - if selected_thread:
+//   - Invalidated mbarrier for all stages
+//
+// Detailed Pseudo-Code:
+// Pre-Prologue loop:
+// __shared__ __mbarrier_t barriers[num_stages];
+// __shared__ __mbarrier_token_t tokens[num_stages];
+// if (warp_id == 0 && electSync()()) {
+//   for (int64_t loop_index : irange(stages)) {
+//     mbarrier_init(mbarrier[loop_index], number_of_arrival_threads);
+//   }
+// }
+//
+// Prologue loop:
+// for (int64_t loop_index : irange(stages-1)) {
+//   if (warp_id == 0 && electSync()()) {
+//     tokens[loop_index] =
+//       mbarrier::arriveExpectTx(mbarrier[loop_index], expected_bytes);
+//     for (...) {
+//       cpAsyncBulk(mbarriers[loop_index], ...);
+//     }
```

Comment:
If I have multiple input TVs, and all of them are circular buffered, what would the code looks like?
```C++
//     tokens[loop_index] =
//       mbarrier::arriveExpectTx(mbarrier[loop_index], expected_bytes);
//     for (...) {
//       cpAsyncBulk1(mbarriers[loop_index], ...);
//     }
//     tokens[loop_index] =
//       mbarrier::arriveExpectTx(mbarrier[loop_index], expected_bytes);
//     for (...) {
//       cpAsyncBulk2(mbarriers[loop_index], ...);
//     }
```
or
```C++
//     tokens[loop_index] =
//       mbarrier::arriveExpectTx(mbarrier[loop_index], expected_bytes1 + expected_bytes2);
//     for (...) {
//       cpAsyncBulk1(mbarriers[loop_index], ...);
//     }
//     for (...) {
//       cpAsyncBulk2(mbarriers[loop_index], ...);
//     }
```
?

Reply:
The first code.

There is a mbarrier token for each TensorView and each circular buffer stage.

Reply:
We might want to merge them to use a single mbarrier in the future, but for now, let's just leave it as is.